File: app/src/main/java/com/zhan_dui/data/MemoEntity.java
package com.zhan_dui.data;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;

/**
 * Room entity for the Memo table.
 */
@Entity(tableName = "Memo")
public class MemoEntity {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "_id")
    private int id;

    @ColumnInfo(name = "content")
    private String content;

    @ColumnInfo(name = "createdtime")
    private long createdTime;

    @ColumnInfo(name = "updatedtime")
    private long updatedTime;

    @ColumnInfo(name = "hash")
    private byte[] hash;

    @ColumnInfo(name = "guid")
    private String guid;

    @ColumnInfo(name = "enid")
    private String enid;

    @ColumnInfo(name = "syncstatus")
    private int syncStatus;

    @ColumnInfo(name = "status")
    private String status;

    @ColumnInfo(name = "cursorposition")
    private int cursorPosition;

    @ColumnInfo(name = "wallid")
    private int wallId;

    @ColumnInfo(name = "orderid")
    private int order;

    @ColumnInfo(name = "lastsynctime")
    private long lastSyncTime;

    @ColumnInfo(name = "attributes")
    private String attributes;

    // Constructors
    public MemoEntity() {
    }

    @Ignore
    public MemoEntity(int id, String content, long createdTime, long updatedTime, byte[] hash,
                     String guid, String enid, int syncStatus, String status, int cursorPosition,
                     int wallId, int order, long lastSyncTime, String attributes) {
        this.id = id;
        this.content = content;
        this.createdTime = createdTime;
        this.updatedTime = updatedTime;
        this.hash = hash;
        this.guid = guid;
        this.enid = enid;
        this.syncStatus = syncStatus;
        this.status = status;
        this.cursorPosition = cursorPosition;
        this.wallId = wallId;
        this.order = order;
        this.lastSyncTime = lastSyncTime;
        this.attributes = attributes;
    }

    // Getters and setters
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public long getCreatedTime() {
        return createdTime;
    }

    public void setCreatedTime(long createdTime) {
        this.createdTime = createdTime;
    }

    public long getUpdatedTime() {
        return updatedTime;
    }

    public void setUpdatedTime(long updatedTime) {
        this.updatedTime = updatedTime;
    }

    public byte[] getHash() {
        return hash;
    }

    public void setHash(byte[] hash) {
        this.hash = hash;
    }

    public String getGuid() {
        return guid;
    }

    public void setGuid(String guid) {
        this.guid = guid;
    }

    public String getEnid() {
        return enid;
    }

    public void setEnid(String enid) {
        this.enid = enid;
    }

    public int getSyncStatus() {
        return syncStatus;
    }

    public void setSyncStatus(int syncStatus) {
        this.syncStatus = syncStatus;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public int getCursorPosition() {
        return cursorPosition;
    }

    public void setCursorPosition(int cursorPosition) {
        this.cursorPosition = cursorPosition;
    }

    public int getWallId() {
        return wallId;
    }

    public void setWallId(int wallId) {
        this.wallId = wallId;
    }

    public int getOrder() {
        return order;
    }

    public void setOrder(int order) {
        this.order = order;
    }

    public long getLastSyncTime() {
        return lastSyncTime;
    }

    public void setLastSyncTime(long lastSyncTime) {
        this.lastSyncTime = lastSyncTime;
    }

    public String getAttributes() {
        return attributes;
    }

    public void setAttributes(String attributes) {
        this.attributes = attributes;
    }

    // Conversion methods to/from Memo (legacy model)
    public static MemoEntity fromMemo(Memo memo) {
        return new MemoEntity(
                memo.getId(),
                memo.getContent(),
                memo.getCreatedTime(),
                memo.getUpdatedTime(),
                memo.getHash(),
                memo.getGuid(),
                memo.getEnid(),
                memo.getSyncStatus(),
                memo.getStatus(),
                memo.getCursorPosition(),
                memo.getWallId(),
                memo.getOrder(),
                memo.getLastSyncTime(),
                memo.getAttributes()
        );
    }

    public Memo toMemo() {
        Memo memo = new Memo();
        memo.setId(id);
        memo.setContent(content);
        memo.setCreatedTime(createdTime);
        memo.setUpdatedTime(updatedTime);
        memo.setHash(hash);
        memo.setGuid(guid);
        memo.setEnid(enid);
        memo.setSyncStatus(syncStatus);
        memo.setStatus(status);
        memo.setCursorPosition(cursorPosition);
        memo.setWallId(wallId);
        memo.setOrder(order);
        memo.setLastSyncTime(lastSyncTime);
        memo.setAttributes(attributes);
        return memo;
    }
}

File: app/src/main/java/com/zhan_dui/data/MemoEntity.kt
package com.zhan_dui.data

import android.os.Parcelable
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.evernote.client.android.EvernoteUtil
import com.evernote.edam.type.Note
import com.zhan_dui.evermemo.Constants
import com.zhan_dui.utils.Logger
import kotlinx.parcelize.Parcelize

/**
 * Room entity for the Memo table.
 */
@Entity(tableName = "Memo")
@Parcelize
data class MemoEntity(
    @JvmField
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "_id")
    var _id: Long? = null,
    @JvmField
    @ColumnInfo(name = "guid")
    var guid: String? = null,

    @JvmField
    @ColumnInfo(name = "enid")
    var enid: String? = null,

    @JvmField
    @ColumnInfo(name = "wallid")
    var wallId: Int = 0,

    @JvmField
    @ColumnInfo(name = "orderid")
    var order: Int = 0,

    @JvmField
    @ColumnInfo(name = "content")
    var content: String? = "",

    @JvmField
    @ColumnInfo(name = "attributes")
    var attributes: String? = null,

    @ColumnInfo(name = "status")
    var status: String? = null,

    @JvmField
    @ColumnInfo(name = "hash")
    var hash: ByteArray? = null,

    @JvmField
    @ColumnInfo(name = "syncstatus")
    var syncStatus: Int = 0,

    @JvmField
    @ColumnInfo(name = "createdtime")
    var createdTime: Long = System.currentTimeMillis(),

    @JvmField
    @ColumnInfo(name = "updatedtime")
    var updatedTime: Long = System.currentTimeMillis(),

    @JvmField
    @ColumnInfo(name = "lastsynctime")
    var lastSyncTime: Long = 0,

    @ColumnInfo(name = "cursorposition")
    var cursorPosition: Int = 0
) : Parcelable {
    fun toNote(notebookGuid: String?): Note {
        val note = toNote()
        note.notebookGuid = notebookGuid
        return note
    }

    fun toNote(): Note {
        return Note().apply {
            title = Constants.NOTEBOOK_TITLE
            content = convertContentToEvernote()
        }
    }

    fun toUpdateNote(): Note {
        val note = toNote()
        note.guid = enid
        return note
    }

    fun toDeleteNote(): Note {
        val note = Note()
        note.guid = enid
        return note
    }

    private fun convertContentToEvernote(): String {
//        val evernoteContent = (EvernoteUtil.NOTE_PREFIX
//                + content?.replace("<br>", "<br/>")
//                + EvernoteUtil.NOTE_SUFFIX)
        val evernoteContent = content?.replace("<br>", "<br/>") ?: ""
        Logger.e(LogTag, "同步文字:$evernoteContent")
        return evernoteContent
    }

    fun setContent(content: String?): MemoEntity {
        this.content = content
        Logger.e("下载下来的文本：$content")
        this.status = STATUS_COMMON
        this.order = 0
        this.attributes = ""
        return this
    }

    fun setNeedSyncDelete() {
        syncStatus = NEED_SYNC_DELETE
    }

    fun setNeedSyncUp() {
        syncStatus = NEED_SYNC_UP
    }

    fun isNeedSyncUp(): Boolean {
        if (syncStatus == NEED_SYNC_UP) {
            return true
        } else {
            return false
        }
    }

    fun isSyncingUp(): Boolean {
        if (syncStatus == SYNCING_UP) {
            return true
        } else {
            return false
        }
    }

    fun isNeedSyncDelete(): Boolean {
        if (syncStatus == NEED_SYNC_DELETE) {
            return true
        } else {
            return false
        }
    }

    fun isDeleted(): Boolean {
        if (status == STATUS_DELETE) {
            return true
        } else {
            return false
        }
    }

    fun buildFromNote(note: Note) {
        content = note.content
        hash = note.contentHash
        updatedTime = note.updated
        createdTime = note.created
        enid = note.guid
        syncStatus = NEED_NOTHING
        status = STATUS_COMMON
        cursorPosition = 0
    }

    companion object {
        const val STATUS_DELETE: String = "delete"
        const val STATUS_COMMON: String = "common"

        private const val LogTag = "Memo"

        /**
         * need to do nothing
         */
        const val NEED_NOTHING: Int = 0

        /**
         * need to sync in Evernote
         */
        const val NEED_SYNC_UP: Int = 1

        /**
         * need to delete in Evernote
         */
        const val NEED_SYNC_DELETE: Int = 3

        /**
         * syning up
         */
        const val SYNCING_UP: Int = 4

        /**
         * syning down
         */
        const val SYNCING_DOWN: Int = 5

        @JvmStatic
        fun buildInsertMemoFromNote(note: Note): MemoEntity {
            return MemoEntity(
                content = note.content,
                hash = note.contentHash,
                updatedTime = note.updated,
                createdTime = note.created,
                enid = note.guid,
                syncStatus = NEED_NOTHING,
                status = STATUS_COMMON,
                cursorPosition = 0,
            )
        }

        @JvmStatic
        fun buildUpdateMemoFromNote(note: Note, id: Long): MemoEntity {
            return buildInsertMemoFromNote(note).apply {
                this._id = id
            }
        }
    }
}